### PR TITLE
Added support for thread exit callbacks

### PR DESCRIPTION
**Pull request by justinboswell**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/posix/thread.c [hunk 1]
```diff
@@ -25,17 +25,32 @@ static struct aws_thread_options s_default_options = {
...
     void *arg;
+    struct thread_atexit_callback *atexit;
 };
 
-static void *thread_fn(void *arg) {
-    struct thread_wrapper wrapper = *(struct thread_wrapper *)arg;
-    aws_mem_release(wrapper.allocator, arg);
+static AWS_THREAD_LOCAL struct thread_wrapper *tl_wrapper = NULL;
 
-    wrapper.func(wrapper.arg);
```
Comment (justinboswell) [inline source/posix/thread.c]: How we ever got past the sanitizer with this, I'll never understand

File: include/aws/common/thread.h [hunk 2]
```diff
@@ -121,6 +121,17 @@ uint64_t aws_thread_current_thread_id(void);
 AWS_COMMON_API
 void aws_thread_current_sleep(uint64_t nanos);
 
+typedef void(aws_thread_atexit_fn)(void);
```
Comment (ColdenCullen) [inline include/aws/common/thread.h]: How about a userdatum?

File: tests/thread_test.c [hunk 3]
```diff
@@ -53,3 +53,37 @@ static int s_test_thread_creation_join_fn(struct aws_allocator *allocator, void
...
+
+static uint32_t s_atexit_call_count = 0;
+static void s_thread_atexit_fn(void *user_data) {
+    (void)user_data;
+    s_atexit_call_count = 1;
+}
+
+static void s_thread_atexit_fn2(void *user_data) {
+    (void)user_data;
+    s_atexit_call_count = 2;
```
Comment (ColdenCullen) [inline tests/thread_test.c]: `AWS_FATAL_ASSERT(s_atexit_call_count == 1);`? Otherwise, if chaining breaks, and we just call the first fn, we're broke.

Comment (justinboswell) [reply tests/thread_test.c]: You are full of good ideas today

File: include/aws/common/thread.h [hunk 4]
```diff
@@ -121,6 +121,17 @@ uint64_t aws_thread_current_thread_id(void);
...
+typedef void(aws_thread_atexit_fn)(void *user_data);
+
+/**
+ * Adds a callback to the chain to be called when the current thread joins.
+ * Callbacks are called from the current thread, in the reverse order they
+ * were added, after the thread function returns.
+ * If not called from within an aws_thread, has no effect.
+ */
+AWS_COMMON_API
+void aws_thread_current_atexit(aws_thread_atexit_fn *callback, void *user_data);
```
Comment (JonathanHenson) [inline include/aws/common/thread.h]: should prolly have an underscore between at and exit

File: source/posix/thread.c [hunk 5]
```diff
@@ -179,3 +196,15 @@ void aws_thread_current_sleep(uint64_t nanos) {
 
     nanosleep(&tm, &output);
 }
+
+void aws_thread_current_atexit(aws_thread_atexit_fn *callback, void *user_data) {
```
Comment (JonathanHenson) [inline source/posix/thread.c]: I suspect a user would want to know this failed.

File: source/posix/thread.c [hunk 6]
```diff
@@ -179,3 +196,15 @@ void aws_thread_current_sleep(uint64_t nanos) {
...
+    if (!tl_wrapper) {
+        return;
+    }
+
+    struct thread_atexit_callback *cb = aws_mem_calloc(tl_wrapper->allocator, 1, sizeof(struct thread_atexit_callback));
```
Comment (JonathanHenson) [inline source/posix/thread.c]: NULL check?

File: source/posix/thread.c [hunk 7]
```diff
@@ -179,3 +196,19 @@ void aws_thread_current_sleep(uint64_t nanos) {
 
     nanosleep(&tm, &output);
 }
+
+int aws_thread_current_at_exit(aws_thread_atexit_fn *callback, void *user_data) {
+    if (!tl_wrapper) {
+        return AWS_OP_ERR;
```
Comment (JonathanHenson) [inline source/posix/thread.c]: shouldn't we raise AWS_ERROR_THREAD_NOT_JOINABLE or something?

Comment (justinboswell) [reply source/posix/thread.c]: Sure!

File: source/posix/thread.c [hunk 8]
```diff
@@ -101,6 +135,7 @@ int aws_thread_launch(
     wrapper->allocator = thread->allocator;
     wrapper->func = func;
     wrapper->arg = arg;
+    wrapper->atexit = NULL;
```
Comment (ColdenCullen) [inline source/posix/thread.c]: Now that you're adding stuff, why not nuke this and switch the malloc to a calloc?

Comment (justinboswell) [reply source/posix/thread.c]: ok

**Pull request merged**